Readme.md

```diff
@@ -0,0 +1,1432 @@
+  [cake-pattern]: https://github.com/alexandru/scala-best-practices/blob/master/sections/3-architecture.md#31-should-not-use-the-cake-pattern
+  
+## Type classes
+
+Type classes are a pattern which relies on implicits resolution to inject pieces of functionality based on one or more types. A classic example of a type class is `Ordering`:
+
+```scala
+trait Ordering[-T] {
+  def compare(left: T, right: T): Int
+}
+
+object Ordering {
+  implicit object IntOrdering extends Ordering[Int] {
+    override def compare(left: Int, right: Int): Int = left - right
+  }
+}
+
+def sorted(v: Vector[T])(implicit ord: Ordering[T]): Vector[T] = {
+  ...
+}
+
+// IntOrdering is injected automatically as
+// the implicit parameter for `sorted`
+val s = sorted(Vector(4, 2, 7, 1, 3))
+```
+
+Type classes are very similar to regular interfaces/traits but allow adding "implementations" to types without changing their `extends` clause. 
+
+So, type classes are a very powerful tool, but, as usual, if they are overused, they can lead to very confusing code. The need to declare custom type classes should be exceedingly rare, therefore unless they significantly increase the clarity of the architecture, they should not be used - regular object-oriented features like traits should be used instead.
+
+However, many important libraries, in particular, for serialization, provide their type classes which are expected to be implemented for custom types. This is usually unavoidable and in some cases lead to reduction of boilerplate code. For example, instead of writing custom low-level serializers for a class to JSON representation, using a predefined method to generate the corresponding type class instance will lead to more succinct and understandable code.
+
+Therefore, you may write such implementations. See the section on implicit values for more information on how to handle them.
+
+
+## Error handling
+
+Exceptions are ubiquitous in the JVM ecosystem, however, by their nature they may make control flow less understandable and clear because they are essentially nonlocal returns - they bubble up the call stack without explicit return statements.
+
+Therefore, basing the control flow on exceptions is discouraged. Not only it makes the code less obvious, it also has performance implications - stack unwinding is not free. Do not use exceptions as a substitute for regular control flow instructions.
+
+### Encoding errors in types
+
+Exceptions should be used, as their name states, in exceptional situations, i.e. when something went wrong and the program cannot continue further because of a broken invariant. This means that you should not use exceptions when it is possible for an error to occur and you want the user of your code to handle it. In such cases errors should be encoded explicitly, using `Option`, `Try`, `Future` or `Either`-like types.
+
+All rules described in the section on using chained method calls apply to the `Try` values, of course.
+
+## Functional programming
```

Comment:
I don't have a problem with how the document handles mutability. I'm using it as an example of how something "non-functional" being used on "small scale" is actually a welcome practice primarily because of the "small scale" part. Because of that, I find the following quote misleading

>Therefore, on the "small scale", i.e. when writing concrete code, it is recommended to follow the functional programming style.

Since both "small-scale" and "functional" are so loosely defined, I'd avoid this recommendation altogether.